### ISSUE TITLE
no-empty-function false positive

### ISSUE THREAD

**Issue by EvgenyOrekhov** [eslint/eslint#12768]
**Tell us about your environment**

* **ESLint Version:** 6.8.0
* **Node Version:** 12.14.1
* **npm Version:** 6.13.4

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```json
{
  "no-empty-function": ["error", { "allow": ["methods"] }]
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const object = {
  async method() {}
};
```

```bash
eslint .
```

**What did you expect to happen?**

No `no-empty-function` warnings.

**What actually happened? Please include the actual, raw output from ESLint.**

> Unexpected empty async method 'method'. (no-empty-function)

**Are you willing to submit a pull request to fix this bug?**

Yes

**Comment by EvgenyOrekhov** [comment]
[ESLint demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyplc2xpbnRcbiAgbm8tZW1wdHktZnVuY3Rpb246IFtcImVycm9yXCIsIHsgXCJhbGxvd1wiOiBbXCJtZXRob2RzXCJdIH1dXG4qL1xuXG5jb25zdCBvYmplY3QgPSB7XG4gIGFzeW5jIG1ldGhvZCgpIHt9XG59O1xuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUta2V5cyI6Miwibm8tZHVwbGljYXRlLWNhc2UiOjIsIm5vLWVtcHR5IjoyLCJuby1lbXB0eS1jaGFyYWN0ZXItY2xhc3MiOjIsIm5vLWVtcHR5LXBhdHRlcm4iOjIsIm5vLWV4LWFzc2lnbiI6Miwibm8tZXh0cmEtYm9vbGVhbi1jYXN0IjoyLCJuby1leHRyYS1zZW1pIjoyLCJuby1mYWxsdGhyb3VnaCI6Miwibm8tZnVuYy1hc3NpZ24iOjIsIm5vLWdsb2JhbC1hc3NpZ24iOjIsIm5vLWlubmVyLWRlY2xhcmF0aW9ucyI6Miwibm8taW52YWxpZC1yZWdleHAiOjIsIm5vLWlycmVndWxhci13aGl0ZXNwYWNlIjoyLCJuby1taXNsZWFkaW5nLWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tbWl4ZWQtc3BhY2VzLWFuZC10YWJzIjoyLCJuby1uZXctc3ltYm9sIjoyLCJuby1vYmotY2FsbHMiOjIsIm5vLW9jdGFsIjoyLCJuby1wcm90b3R5cGUtYnVpbHRpbnMiOjIsIm5vLXJlZGVjbGFyZSI6Miwibm8tcmVnZXgtc3BhY2VzIjoyLCJuby1zZWxmLWFzc2lnbiI6Miwibm8tc2hhZG93LXJlc3RyaWN0ZWQtbmFtZXMiOjIsIm5vLXNwYXJzZS1hcnJheXMiOjIsIm5vLXRoaXMtYmVmb3JlLXN1cGVyIjoyLCJuby11bmRlZiI6Miwibm8tdW5leHBlY3RlZC1tdWx0aWxpbmUiOjIsIm5vLXVucmVhY2hhYmxlIjoyLCJuby11bnNhZmUtZmluYWxseSI6Miwibm8tdW5zYWZlLW5lZ2F0aW9uIjoyLCJuby11bnVzZWQtbGFiZWxzIjoyLCJuby11bnVzZWQtdmFycyI6Miwibm8tdXNlbGVzcy1jYXRjaCI6Miwibm8tdXNlbGVzcy1lc2NhcGUiOjIsIm5vLXdpdGgiOjIsInJlcXVpcmUtYXRvbWljLXVwZGF0ZXMiOjIsInJlcXVpcmUteWllbGQiOjIsInVzZS1pc25hbiI6MiwidmFsaWQtdHlwZW9mIjoyfSwiZW52Ijp7fX19)

**Comment by EvgenyOrekhov** [comment]
I can see in the [source code](https://github.com/eslint/eslint/blob/a1d999c6b4e51c317ad409110be7be214ff9f7c6/lib/rules/no-empty-function.js) that there is actually some code for handling async methods and functions. But it's basically dead code now. I guess one of the possible solutions would be to extend the [ALLOW_OPTIONS](https://github.com/eslint/eslint/blob/a1d999c6b4e51c317ad409110be7be214ff9f7c6/lib/rules/no-empty-function.js#L18) array to include "asyncFunctions" and "asyncMethods". That way it would be an enhancement, I guess.

**Comment by kaicataldo** [comment]
Verified using [our demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50ICBcIm5vLWVtcHR5LWZ1bmN0aW9uXCI6IFtcImVycm9yXCIsIHsgXCJhbGxvd1wiOiBbXCJtZXRob2RzXCJdIH1dICovXG5cbmNvbnN0IG9iamVjdCA9IHtcbiAgYXN5bmMgbWV0aG9kKCkge31cbn07Iiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMSwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnsiYnJvd3NlciI6ZmFsc2V9fX0=).

Given that we have a separate option for `"generatorFunctions"` and `"generatorMethods"`, I personally think it makes sense to add `"asyncFunctions"` and `"asyncMethods"` as an option.

I'm not sure whether to mark this as a bug or an enhancement, since adding new options is an enhancement, but given the lack of these options, I think this should also be classified as a bug. 🤷‍♂ 

Either way, I support solving this by adding `"allow": "asyncFunctions|asyncMethods"` options. Let' see what the rest of the team has to say.

**Comment by mdjermanovic** [comment]
I also think this should be fixed in some way. There are also async arrow functions.

By the current implementation, it seems that async generator functions and async generator methods would be treated as just generator functions/generator methods? Looks like there could be an exponential growth of options if something new appears.

**Comment by kaicataldo** [comment]
On thinking about this more, this seems like a clear case of an unintentionally missing feature (it was probably implemented before async functions were added to the language). The rule definitely doesn't feel complete without it, so I'm marking as accepted. If anyone from the team disagrees, please let me know!
